pykern/pkcli/ci.py

```diff
@@ -8,59 +8,53 @@
 :copyright: Copyright (c) 2022 RadiaSoft LLC.  All Rights Reserved.
 :license: http://www.apache.org/licenses/LICENSE-2.0.html
 """
-from pykern.pkdebug import pkdp, pkdlog
-from pykern import pkunit
-from pykern.pkcli import test
+from pykern import pkcli
+from pykern import pkio
 from pykern import pksetup
+from pykern import pkunit
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdp, pkdlog
 import re
 
-
-_FILE_TYPE = re.compile(r".py$")
-_EXCLUDE_FILES = re.compile(
+_CHECK_EOF_NEWLINE_EXCLUDE_FILES = re.compile(
+    r"/ext/|node_modules/|^run/|tests/|^venv/"
+)
+_CHECK_EOF_NEWLINE_INCLUDE_FILES = re.compile(r"\.(html|jinja|js|json|md|py|tsx|yml)$")
+_CHECK_PRINTS_EXCLUDE_FILES = re.compile(
     f".*(?:{pkunit.DATA_DIR_SUFFIX}|{pkunit.WORK_DIR_SUFFIX})/"
     + f"|^\\w+/{pksetup.PACKAGE_DATA}/"
     + r"|pkdebug[^/]*\.py$"
     + r"|(?:^|/)\."
     + r"|^run/"
+    + r"|^venv/"
 )
+_CHECK_PRINTS_INCLUDE_FILES = re.compile(r".py$")
 _PRINT = re.compile(r"(?:\s|^)(?:pkdp|print)\(")
 _PRINT_OK = re.compile(r"^\s*#\s*(?:pkdp|print)\(")
 
 
+def check_eof_newline():
+    """Recursively check repo for files missing newline at end of file.
+
+    Files matching _CHECK_EOF_NEWLINE_FILE_EXTS and not matching
+    _CHECK_EOF_NEWLINE_EXCLUDE_FILES will be checked.
+    """
+    _check_files("check_eof_newline")
```

Comment:
You could simplify the code and remove dispatch by calling _check_files this way:

```py
def check_eof_newline():
    def _file(lines):
        return None if lines[-1] == "" else ""

    return _check_files("check_eof_newline", _file)
```

None means no error. Have _check_files prefix the file_path to the result and append to res so you don't have to pass either. Both routines need lines so share that code.

Reply:
16e58b893588afe93755e152fe7397ee18ff7eaf